File: app/src/main/java/com/demo/app/ncov2020/MainActivity.java
package com.demo.app.ncov2020;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Color;
import android.os.Bundle;
import com.mapbox.mapboxsdk.Mapbox;
import com.mapbox.mapboxsdk.annotations.MarkerOptions;
import com.mapbox.mapboxsdk.annotations.PolygonOptions;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.geometry.LatLngBounds;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.mapboxsdk.offline.OfflineManager;
import com.mapbox.mapboxsdk.offline.OfflineRegion;
import com.mapbox.mapboxsdk.offline.OfflineRegionError;
import com.mapbox.mapboxsdk.offline.OfflineRegionStatus;
import com.mapbox.mapboxsdk.offline.OfflineTilePyramidRegionDefinition;
import com.mapbox.mapboxsdk.style.layers.Property;
import com.mapbox.mapboxsdk.style.layers.SymbolLayer;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ProgressBar;
import android.widget.Toast;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconAnchor;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconImage;

public class MainActivity extends AppCompatActivity {
    private MapView mapView;


    private boolean isEndNotified;
    private ProgressBar progressBar;

    private OfflineManager offlineManager;

    // JSON encoding/decoding
    public static final String JSON_CHARSET = "UTF-8";
    public static final String JSON_FIELD_REGION_NAME = "FIELD_REGION_NAME";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        this.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,WindowManager.LayoutParams.FLAG_FULLSCREEN);

        Mapbox.getInstance(this, "pk.eyJ1IjoibmNvdmdhbWUiLCJhIjoiY2s3eWpjcjJjMDdnZTNqcGZ2ZXBxMGYxdSJ9.IBqgc27bmXnxY2G6iF-MiQ");
        setContentView(R.layout.activity_main);

        mapView = findViewById(R.id.mapView);
        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(@NonNull final MapboxMap mapboxMap) {
                mapboxMap.setMaxZoomPreference(2);
                mapboxMap.setMinZoomPreference(1);
                mapboxMap.getUiSettings().setRotateGesturesEnabled(false);
                mapboxMap.addMarker(new MarkerOptions()
                        .position(new LatLng(48.85819, 2.29458))
                        .title("Eiffel Tower"));
                List<LatLng> polygonLatLngList = new ArrayList<>();

                polygonLatLngList.add(new LatLng(45.522585, 122.685699));
                polygonLatLngList.add(new LatLng(45.534611, -122.708873));
                polygonLatLngList.add(new LatLng(45.530883, -122.678833));
                polygonLatLngList.add(new LatLng(45.547115, -122.667503));
                polygonLatLngList.add(new LatLng(45.530643, -122.660121));
                polygonLatLngList.add(new LatLng(45.533529, -122.636260));
                polygonLatLngList.add(new LatLng(45.521743, -122.659091));
                polygonLatLngList.add(new LatLng(45.510677, -122.648792));
                polygonLatLngList.add(new LatLng(45.515008, -122.664070));
                polygonLatLngList.add(new LatLng(45.502496, -122.669048));
                polygonLatLngList.add(new LatLng(45.515369, -122.678489));
                polygonLatLngList.add(new LatLng(45.506346, -122.702007));
                polygonLatLngList.add(new LatLng(45.522585, 122.685699));

                mapboxMap.addPolygon(new PolygonOptions()
                        .addAll(polygonLatLngList)
                        .fillColor(Color.parseColor("#3bb2d0")));
                mapboxMap.setStyle(Style.DARK, new Style.OnStyleLoaded() {
                    @Override
                    public void onStyleLoaded(@NonNull Style style) {

// Set up the OfflineManager
                        offlineManager = OfflineManager.getInstance(MainActivity.this);

// Create a bounding box for the offline region
                        LatLngBounds latLngBounds = new LatLngBounds.Builder()
                                .include(new LatLng(37.7897, -119.5073)) // Northeast
                                .include(new LatLng(37.6744, -119.6815)) // Southwest
                                .build();

// Define the offline region
                        OfflineTilePyramidRegionDefinition definition = new OfflineTilePyramidRegionDefinition(
                                style.getUri(),
                                latLngBounds,
                                10,
                                20,
                                MainActivity.this.getResources().getDisplayMetrics().density);

// Set the metadata
                        byte[] metadata;
                        try {
                            JSONObject jsonObject = new JSONObject();
                            jsonObject.put(JSON_FIELD_REGION_NAME, "Yosemite National Park");
                            String json = jsonObject.toString();
                            metadata = json.getBytes(JSON_CHARSET);
                        } catch (Exception exception) {
                            Timber.e("Failed to encode metadata: %s", exception.getMessage());
                            metadata = null;
                        }

// Create the region asynchronously
                        if (metadata != null) {
                            offlineManager.createOfflineRegion(
                                    definition,
                                    metadata,
                                    new OfflineManager.CreateOfflineRegionCallback() {
                                        @Override
                                        public void onCreate(OfflineRegion offlineRegion) {
                                            offlineRegion.setDownloadState(OfflineRegion.STATE_ACTIVE);

// Display the download progress bar
                                            progressBar = findViewById(R.id.progress_bar);
                                            startProgress();

// Monitor the download progress using setObserver
                                            offlineRegion.setObserver(new OfflineRegion.OfflineRegionObserver() {
                                                @Override
                                                public void onStatusChanged(OfflineRegionStatus status) {

// Calculate the download percentage and update the progress bar
                                                    double percentage = status.getRequiredResourceCount() >= 0
                                                            ? (100.0 * status.getCompletedResourceCount() / status.getRequiredResourceCount()) :
                                                            0.0;

                                                    if (status.isComplete()) {
// Download complete
                                                        endProgress("End process");
                                                    } else if (status.isRequiredResourceCountPrecise()) {
// Switch to determinate state
                                                        setPercentage((int) Math.round(percentage));
                                                    }
                                                }

                                                @Override
                                                public void onError(OfflineRegionError error) {
// If an error occurs, print to logcat
                                                    Timber.e("onError reason: %s", error.getReason());
                                                    Timber.e("onError message: %s", error.getMessage());
                                                }

                                                @Override
                                                public void mapboxTileCountLimitExceeded(long limit) {
// Notify if offline region exceeds maximum tile count
                                                    Timber.e("Mapbox tile count limit exceeded: %s", limit);
                                                }
                                            });
                                        }

                                        @Override
                                        public void onError(String error) {
                                            Timber.e("Error: %s", error);
                                        }
                                    });
                        }
                    }
                });
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        mapView.onResume();
    }

    @Override
    protected void onStart() {
        super.onStart();
        mapView.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mapView.onStop();
    }

    @Override
    public void onPause() {
        super.onPause();
        mapView.onPause();
        if (offlineManager != null) {
            offlineManager.listOfflineRegions(new OfflineManager.ListOfflineRegionsCallback() {
                @Override
                public void onList(OfflineRegion[] offlineRegions) {
                    if (offlineRegions.length > 0) {
// delete the last item in the offlineRegions list which will be yosemite offline map
                        offlineRegions[(offlineRegions.length - 1)].delete(new OfflineRegion.OfflineRegionDeleteCallback() {
                            @Override
                            public void onDelete() {
                                Toast.makeText(
                                        MainActivity.this,
                                        "Deleted toasts",
                                        Toast.LENGTH_LONG
                                ).show();
                            }

                            @Override
                            public void onError(String error) {
                                Timber.e("On delete error: %s", error);
                            }
                        });
                    }
                }

                @Override
                public void onError(String error) {
                    Timber.e("onListError: %s", error);
                }
            });
        }
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mapView.onSaveInstanceState(outState);
    }

    // Progress bar methods
    private void startProgress() {

// Start and show the progress bar
        isEndNotified = false;
        progressBar.setIndeterminate(true);
        progressBar.setVisibility(View.VISIBLE);
    }

    private void setPercentage(final int percentage) {
        progressBar.setIndeterminate(false);
        progressBar.setProgress(percentage);
    }

    private void endProgress(final String message) {
// Don't notify more than once
        if (isEndNotified) {
            return;
        }

// Stop and hide the progress bar
        isEndNotified = true;
        progressBar.setIndeterminate(false);
        progressBar.setVisibility(View.GONE);

// Show a toast
        Toast.makeText(MainActivity.this, message, Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/com/demo/app/ncov2020/MainActivity.kt
package com.demo.app.ncov2020

import android.os.Bundle
import android.view.Window
import android.view.WindowManager
import androidx.navigation.Navigation
import com.demo.app.basics.mvvm.BaseActivity
import com.demo.app.ncov2020.common.GameNavigatorImpl
import com.demo.app.ncov2020.common.GameNavigatorImpl.Companion.instance

class MainActivity : BaseActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        setContentView(R.layout.activity_main)
        //getSupportFragmentManager().beginTransaction().replace(R.id.container, new LoginFragment()).commit();
        instance.setNavController(Navigation.findNavController(this, R.id.nav_host_fragment))
    }

    override fun onDestroy() {
        instance.releaseNavController()
        super.onDestroy()
    } //    @Override
//    public void onResume() {
//        super.onResume();
//        mapView.onResume();
//    }
//
//    @Override
//    protected void onStart() {
//        super.onStart();
//        mapView.onStart();
//    }
//
//    @Override
//    protected void onStop() {
//        super.onStop();
//        mapView.onStop();
//    }
//
//    @Override
//    public void onPause() {
//        super.onPause();
//        mapView.onPause();
////        if (offlineManager != null) {
////            offlineManager.listOfflineRegions(new OfflineManager.ListOfflineRegionsCallback() {
////                @Override
////                public void onList(OfflineRegion[] offlineRegions) {
////                    if (offlineRegions.length > 0) {
////// delete the last item in the offlineRegions list which will be yosemite offline map
////                        offlineRegions[(offlineRegions.length - 1)].delete(new OfflineRegion.OfflineRegionDeleteCallback() {
////                            @Override
////                            public void onDelete() {
////                                Toast.makeText(
////                                        MainActivity.this,
////                                        "Deleted toasts",
////                                        Toast.LENGTH_LONG
////                                ).show();
////                            }
////
////                            @Override
////                            public void onError(String error) {
////                                Timber.e("On delete error: %s", error);
////                            }
////                        });
////                    }
////                }
////
////                @Override
////                public void onError(String error) {
////                    Timber.e("onListError: %s", error);
////                }
////            });
////        }
//    }
//
//    @Override
//    public void onLowMemory() {
//        super.onLowMemory();
//        mapView.onLowMemory();
//    }
//
//    @Override
//    protected void onDestroy() {
//        super.onDestroy();
//        mapView.onDestroy();
//    }
//
//    @Override
//    protected void onSaveInstanceState(Bundle outState) {
//        super.onSaveInstanceState(outState);
//        mapView.onSaveInstanceState(outState);
//    }
//
//    // Progress bar methods
//    private void startProgress() {
//
//// Start and show the progress bar
//        isEndNotified = false;
//        progressBar.setIndeterminate(true);
//        progressBar.setVisibility(View.VISIBLE);
//    }
//
//    private void setPercentage(final int percentage) {
//        progressBar.setIndeterminate(false);
//        progressBar.setProgress(percentage);
//    }
//
//    private void endProgress(final String message) {
//// Don't notify more than once
//        if (isEndNotified) {
//            return;
//        }
//
//// Stop and hide the progress bar
//        isEndNotified = true;
//        progressBar.setIndeterminate(false);
//        progressBar.setVisibility(View.GONE);
//
//// Show a toast
//        Toast.makeText(MainActivity.this, message, Toast.LENGTH_LONG).show();
//    }
}